app/src/main/java/nz/org/cacophony/birdmonitor/TestRecordFragment.java

```diff
@@ -177,4 +192,124 @@ public void onReceive(Context context, Intent intent) {
+
+        return allPermissionsAlreadyGranted;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String[] permissions,
+                                           int[] grantResults) {
+        // BEGIN_INCLUDE(onRequestPermissionsResult)
+        if (requestCode == PERMISSION_WRITE_EXTERNAL_STORAGE) {
+            // Request for camera permission.
+            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // Permission has been granted. Start recording
+                tvMessages.setText("WRITE_EXTERNAL_STORAGE permission granted");
+            } else {
+                tvMessages.setText("Do not have WRITE_EXTERNAL_STORAGE permission, You can NOT save recordings");
+            }
+        }
+
+        if (requestCode == PERMISSION_RECORD_AUDIO) {
+            // Request for camera permission.
+            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // Permission has been granted. Start recording
+                tvMessages.setText("RECORD_AUDIO permission granted");
+            } else {
+                tvMessages.setText("Do not have RECORD_AUDIO permission, You can NOT record");
+            }
+        }
+
+        // May as well check GPS Location permission again
+
+        if (requestCode == PERMISSION_LOCATION) {
+            // Request for camera permission.
+            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // Permission has been granted. Start recording
+                tvMessages.setText("LOCATION permission granted");
+            } else {
+                tvMessages.setText("Do not have LOCATION permission, You can NOT set the GPS position");
+            }
+        }
+
+        // To avoid user having to press the 'Record Now' button again, we will check to see if we know have all permissions and if we do, start the recording
+
+        if (haveAllPermissions(getActivity().getApplicationContext())){
+            recordNow();
+        }
+
+        // END_INCLUDE(onRequestPermissionsResult)
+    }
+
+    private boolean haveAllPermissions(Context context){
+        boolean allPermissionsAlreadyGranted = true;
+
+        if (ContextCompat.checkSelfPermission(context,
+                Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            allPermissionsAlreadyGranted = false;
+
+        }
+
+        if (ContextCompat.checkSelfPermission(context,
+                Manifest.permission.RECORD_AUDIO)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            allPermissionsAlreadyGranted = false;
+
+        }
+
+        if (ContextCompat.checkSelfPermission(context,
+                Manifest.permission.ACCESS_FINE_LOCATION)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            allPermissionsAlreadyGranted = false;
+
+        }
```

Comment:
Instead of repeating this code 3 times how about looping over an array of permissions to check?

Reply:
ok